prov/rxm/src/rxm_ep.c

```diff
@@ -1068,34 +1100,58 @@ static int rxm_ep_msg_res_open(struct fi_info *rxm_fi_info,
 	rxm_domain = container_of(util_domain, struct rxm_domain, util_domain);
 	rxm_fabric = container_of(util_domain->fabric, struct rxm_fabric, util_fabric);
 
-	ret = fi_passive_ep(rxm_fabric->msg_fabric, rxm_ep->msg_info, &rxm_ep->msg_pep, rxm_ep);
+	eq_attr.wait_obj = FI_WAIT_UNSPEC;
+	eq_attr.flags = FI_WRITE;
+
+	ret = fi_eq_open(rxm_fabric->msg_fabric, &eq_attr, &rxm_ep->msg_eq, NULL);
 	if (ret) {
-		FI_WARN(&rxm_prov, FI_LOG_FABRIC, "Unable to open msg PEP\n");
+		FI_WARN(&rxm_prov, FI_LOG_FABRIC, "Unable to open msg EQ\n");
 		goto err1;
 	}
 
+	if (pthread_create(&rxm_ep->msg_listener_thread, 0,
+			   rxm_msg_listener, rxm_ep)) {
+		ret = -errno;
+		FI_WARN(&rxm_prov, FI_LOG_FABRIC,
+			"Unable to create msg_cm_listener_thread\n");
+		goto err2;
+	}
+
+	ret = fi_passive_ep(rxm_fabric->msg_fabric, rxm_ep->msg_info,
+			    &rxm_ep->msg_pep, rxm_ep);
+	if (ret) {
+		FI_WARN(&rxm_prov, FI_LOG_FABRIC, "Unable to open msg PEP\n");
+		goto err3;
+	}
+
+	ret = fi_pep_bind(rxm_ep->msg_pep, &rxm_ep->msg_eq->fid, 0);
+	if (ret) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+				"Unable to bind msg PEP to msg EQ\n");
+		goto err4;
+	}
+	ret = fi_listen(rxm_ep->msg_pep);
+	if (ret) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+				"Unable to set msg PEP to listen state\n");
+		goto err4;
+	}
```

Comment:
I would move the listen handling portion into another function to shorten this one and group that functionality together.  Only start the listen thread after everything else has been setup.